pkg/streamingpromql/operators/functions/absent_over_time.go

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/functions.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package functions
+
+import (
+	"context"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// AbsentOverTime performs a rate calculation over a range vector.
+type AbsentOverTime struct {
+	timeRange                types.QueryTimeRange
+	argExpressions           parser.Expr
+	inner                    types.RangeVectorOperator
+	expressionPosition       posrange.PositionRange
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	presence                 []bool
+	exhausted                bool
+}
+
+var _ types.InstantVectorOperator = &AbsentOverTime{}
+
+func NewAbsentOverTime(
+	inner types.RangeVectorOperator,
+	argExpressions parser.Expr,
+	timeRange types.QueryTimeRange,
+	expressionPosition posrange.PositionRange,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+) *AbsentOverTime {
+	return &AbsentOverTime{
+		timeRange:                timeRange,
+		inner:                    inner,
+		argExpressions:           argExpressions,
+		expressionPosition:       expressionPosition,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+	}
+}
+
+func (a *AbsentOverTime) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	innerMetadata, err := a.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer types.PutSeriesMetadataSlice(innerMetadata)
+
+	a.presence, err = types.BoolSlicePool.Get(a.timeRange.StepCount, a.memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize presence slice
+	a.presence = a.presence[:a.timeRange.StepCount]
+
+	metadata := types.GetSeriesMetadataSlice(1)
+	metadata = append(metadata, types.SeriesMetadata{
+		Labels: createLabelsForAbsentFunction(a.argExpressions),
+	})
+
+	for range innerMetadata {
+		err := a.inner.NextSeries(ctx)
+		if err != nil {
+			return nil, err
+		}
+		for {
+			step, err := a.inner.NextStepSamples()
+			// nolint:errorlint // errors.Is introduces a performance overhead, and NextStepSamples is guaranteed to return exactly EOS, never a wrapped error.
+			if err == types.EOS {
+				break
```

Comment:
We could simplify this slightly by changing the `for` to something like `for stepIdx := range a.timeRange.StepCount`. We can then drop the special check for `err == types.EOS` (we should never receive an `EOS` for any of the steps, and if we do, we should return an error) and line 82 then doesn't need to call `PointIndex`.

Reply:
Updated in dcd5239fdf2cf631c87b0f8dcbc8d807cc257e4f